File: Packages/NMPKotlin/src/main/kotlin/com/nmp/sdk/ReceiptSet.kt
package com.nmp.sdk

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import uniffi.nmp_ffi.FfiReceiptIdentity
import uniffi.nmp_ffi.FfiReceiptSetEvent
import uniffi.nmp_ffi.FfiReceiptSetException

sealed interface ReceiptSetIdentity {
    data class Id(val receiptId: ULong) : ReceiptSetIdentity
    data class Correlation(val token: String) : ReceiptSetIdentity

    fun toFfi(): FfiReceiptIdentity =
        when (this) {
            is Id -> FfiReceiptIdentity.Id(receiptId)
            is Correlation -> FfiReceiptIdentity.Correlation(token)
        }
}

sealed interface ReceiptSetEvent {
    data class Fact(
        val identity: ReceiptSetIdentity,
        val receiptId: ULong,
        val status: WriteStatus,
    ) : ReceiptSetEvent

    data class NotFound(val identity: ReceiptSetIdentity) : ReceiptSetEvent
    data class RetainedButUnreadable(
        val identity: ReceiptSetIdentity,
        val receiptId: ULong?,
    ) : ReceiptSetEvent
    data class ReplayAfterLag(
        val identity: ReceiptSetIdentity,
        val receiptId: ULong,
    ) : ReceiptSetEvent
    data class ReplayUnavailable(
        val identity: ReceiptSetIdentity,
        val receiptId: ULong,
    ) : ReceiptSetEvent
    data class Closed(
        val identity: ReceiptSetIdentity,
        val receiptId: ULong,
    ) : ReceiptSetEvent
}

sealed class NMPReceiptSetError(message: String) : Exception(message) {
    data class CapacityExceeded(val capacity: ULong, val requested: ULong) :
        NMPReceiptSetError("receipt set capacity $capacity exceeded by $requested identities")
    data class DuplicateIdentity(val identity: String) :
        NMPReceiptSetError("duplicate receipt identity $identity")
    data object EngineClosed : NMPReceiptSetError("engine already shut down")
}

val NMPEngine.receiptSetCapacity: ULong
    get() = ffi.receiptSetCapacity()

fun NMPEngine.observeReceipts(identities: List<ReceiptSetIdentity>): Flow<ReceiptSetEvent> {
    val stream =
        try {
            ffi.observeReceipts(identities.map(ReceiptSetIdentity::toFfi))
        } catch (error: FfiReceiptSetException) {
            throw when (error) {
                is FfiReceiptSetException.CapacityExceeded ->
                    NMPReceiptSetError.CapacityExceeded(error.capacity, error.requested)
                is FfiReceiptSetException.DuplicateIdentity ->
                    NMPReceiptSetError.DuplicateIdentity(error.identity)
                is FfiReceiptSetException.EngineClosed -> NMPReceiptSetError.EngineClosed
            }
        }
    return flow {
        try {
            while (true) {
                val event = nmpRethrowingAsync { stream.next() } ?: break
                emit(event.toSdk())
            }
        } finally {
            stream.cancel()
        }
    }
}

private fun FfiReceiptIdentity.toSdk(): ReceiptSetIdentity =
    when (this) {
        is FfiReceiptIdentity.Id -> ReceiptSetIdentity.Id(receiptId)
        is FfiReceiptIdentity.Correlation -> ReceiptSetIdentity.Correlation(token)
    }

private fun FfiReceiptSetEvent.toSdk(): ReceiptSetEvent =
    when (this) {
        is FfiReceiptSetEvent.Fact ->
            ReceiptSetEvent.Fact(identity.toSdk(), receiptId, WriteStatus.from(status))
        is FfiReceiptSetEvent.NotFound -> ReceiptSetEvent.NotFound(identity.toSdk())
        is FfiReceiptSetEvent.RetainedButUnreadable ->
            ReceiptSetEvent.RetainedButUnreadable(identity.toSdk(), receiptId)
        is FfiReceiptSetEvent.ReplayAfterLag ->
            ReceiptSetEvent.ReplayAfterLag(identity.toSdk(), receiptId)
        is FfiReceiptSetEvent.ReplayUnavailable ->
            ReceiptSetEvent.ReplayUnavailable(identity.toSdk(), receiptId)
        is FfiReceiptSetEvent.Closed -> ReceiptSetEvent.Closed(identity.toSdk(), receiptId)
    }

File: Packages/NMP/Sources/NMP/ReceiptSet.swift
import NMPFFI

/// One retained receipt identity in a bounded receipt-set observation.
public enum ReceiptSetIdentity: Sendable, Hashable {
    case id(UInt64)
    case correlation(String)

    func toFfi() -> FfiReceiptIdentity {
        switch self {
        case .id(let id): .id(receiptId: id)
        case .correlation(let token): .correlation(token: token)
        }
    }
}

/// Tagged receipt facts and exact per-identity recovery outcomes.
public enum ReceiptSetEvent: Sendable {
    case fact(identity: ReceiptSetIdentity, receiptId: UInt64, status: WriteStatus)
    case notFound(identity: ReceiptSetIdentity)
    case retainedButUnreadable(identity: ReceiptSetIdentity, receiptId: UInt64?)
    case replayAfterLag(identity: ReceiptSetIdentity, receiptId: UInt64)
    case replayUnavailable(identity: ReceiptSetIdentity, receiptId: UInt64)
    /// Emitted only after this receipt's complete replay/live stream closes.
    case closed(identity: ReceiptSetIdentity, receiptId: UInt64)

    init(_ event: FfiReceiptSetEvent) {
        switch event {
        case .fact(let identity, let receiptId, let status):
            self = .fact(
                identity: ReceiptSetIdentity(identity),
                receiptId: receiptId,
                status: WriteStatus(status)
            )
        case .notFound(let identity):
            self = .notFound(identity: ReceiptSetIdentity(identity))
        case .retainedButUnreadable(let identity, let receiptId):
            self = .retainedButUnreadable(
                identity: ReceiptSetIdentity(identity),
                receiptId: receiptId
            )
        case .replayAfterLag(let identity, let receiptId):
            self = .replayAfterLag(
                identity: ReceiptSetIdentity(identity),
                receiptId: receiptId
            )
        case .replayUnavailable(let identity, let receiptId):
            self = .replayUnavailable(
                identity: ReceiptSetIdentity(identity),
                receiptId: receiptId
            )
        case .closed(let identity, let receiptId):
            self = .closed(identity: ReceiptSetIdentity(identity), receiptId: receiptId)
        }
    }
}

private extension ReceiptSetIdentity {
    init(_ identity: FfiReceiptIdentity) {
        switch identity {
        case .id(let receiptId): self = .id(receiptId)
        case .correlation(let token): self = .correlation(token)
        }
    }
}

public enum NMPReceiptSetError: Error, Sendable, Equatable {
    case capacityExceeded(capacity: UInt64, requested: UInt64)
    case duplicateIdentity(String)
    case engineClosed

    init(_ error: FfiReceiptSetError) {
        switch error {
        case .CapacityExceeded(let capacity, let requested):
            self = .capacityExceeded(capacity: capacity, requested: requested)
        case .DuplicateIdentity(let identity):
            self = .duplicateIdentity(identity)
        case .EngineClosed:
            self = .engineClosed
        }
    }
}

/// One fair pull sequence over a finite set of retained receipts.
public struct ReceiptSetStatus: AsyncSequence, Sendable {
    public typealias Element = ReceiptSetEvent

    private let handle: NmpReceiptSetStream
    private let iteratorGate = NMPPullIteratorGate()

    init(handle: NmpReceiptSetStream) {
        self.handle = handle
    }

    public func makeAsyncIterator() -> Iterator {
        Iterator(
            core: NMPPullIteratorCore(
                handle: handle,
                iteratorGate: iteratorGate,
                map: { @Sendable event in ReceiptSetEvent(event) }
            )
        )
    }

    public struct Iterator: AsyncIteratorProtocol {
        let core: NMPPullIteratorCore<NmpReceiptSetStream, ReceiptSetEvent>

        public mutating func next() async throws -> ReceiptSetEvent? {
            try await core.next()
        }
    }

    public func cancel() {
        handle.cancel()
    }
}

extension NMPEngine {
    /// The exact NMP-owned admission bound for one receipt-set observation.
    public var receiptSetCapacity: UInt64 {
        ffi.receiptSetCapacity()
    }

    /// Observe every retained identity fairly through one cancellation scope.
    public func observeReceipts(
        _ identities: [ReceiptSetIdentity]
    ) throws -> ReceiptSetStatus {
        do {
            return ReceiptSetStatus(
                handle: try ffi.observeReceipts(identities: identities.map { $0.toFfi() })
            )
        } catch let error as FfiReceiptSetError {
            throw NMPReceiptSetError(error)
        }
    }
}
